libtenzir/builtins/operators/fork.cpp

```diff
@@ -6,61 +6,242 @@
+        }
+        push_rp_ = self_->make_response_promise<void>();
+        return push_rp_;
+      },
+      [this](atom::pull) -> caf::result<table_slice> {
+        TENZIR_ASSERT(not pull_rp_.pending());
+        if (buffer_.empty()) {
+          pull_rp_ = self_->make_response_promise<table_slice>();
+          return pull_rp_;
+        }
+        if (push_rp_.pending()) {
+          TENZIR_ASSERT(buffer_.size() == max_buffered);
+          push_rp_.deliver();
+        }
+        auto output = std::move(buffer_.front());
+        buffer_.pop();
+        return output;
+      },
+      [this](uint64_t operator_index, uuid metrics_id,
+             type schema) -> caf::result<void> {
+        TENZIR_UNUSED(operator_index);
+        return self_
+          ->mail(parent_operator_index_, metrics_id, std::move(schema))
+          .delegate(metrics_receiver_);
+      },
+      [this](uint64_t operator_index, uuid metrics_id,
+             record metrics) -> caf::result<void> {
+        TENZIR_UNUSED(operator_index);
+        return self_
+          ->mail(parent_operator_index_, metrics_id, std::move(metrics))
+          .delegate(metrics_receiver_);
+      },
+      [](const operator_metric& metrics) -> caf::result<void> {
+        // Operator metrics cannot be forwarded.
+        TENZIR_UNUSED(metrics);
+        return {};
+      },
+      [this](diagnostic diag) -> caf::result<void> {
+        diagnostics_handler_.emit(std::move(diag));
+        return {};
+      },
+    };
+  }
+
+private:
+  static constexpr auto max_buffered = size_t{10};
+  caf::typed_response_promise<void> push_rp_;
+  caf::typed_response_promise<table_slice> pull_rp_;
+  std::queue<table_slice> buffer_;
+
+  side_channel_actor::pointer self_;
+  shared_diagnostic_handler diagnostics_handler_;
+  metrics_receiver_actor metrics_receiver_;
+  uint64_t parent_operator_index_;
+};
+
+class internal_fork_source_operator final
+  : public crtp_operator<internal_fork_source_operator> {
+public:
+  internal_fork_source_operator() = default;
+
+  internal_fork_source_operator(side_channel_actor side_channel)
+    : side_channel_{std::move(side_channel)} {
+  }
+
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // Signal the start immediately as the parent pipeline won't deliver results
+    // before the nested pipeline has started up.
+    co_yield {};
+    auto result = table_slice{};
+    do {
+      ctrl.self()
+        .mail(atom::pull_v)
+        .request(side_channel_, caf::infinite)
+        .then(
+          [&](table_slice output) {
+            result = std::move(output);
+            ctrl.set_waiting(false);
+          },
+          [&](caf::error err) {
+            diagnostic::error(std::move(err))
+              .note("failed to accept forwarded events")
+              .emit(ctrl.diagnostics());
+          });
+      ctrl.set_waiting(true);
+      co_yield std::exchange(result, {});
+    } while (result.rows() > 0);
```

Comment:
This will terminate on the first iteration.

Reply:
No, it doesn't. We suspend before we yield, and then when we get back `result` will no longer be empty.

Reply:
Ah, right.